kythe/rust/indexer/src/providers.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 The Kythe Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::error::KytheError;
+use analysis_rust_proto::*;
+use positioned_io;
+use rc_zip::{prelude::*, Archive, EntryContents};
+use std::fs::File;
+use std::io::Read;
+
+/// A trait for retrieving files during indexing.
+pub trait FileProvider {
```

Comment:
Could this trait be defined in a way that doesn't require every implementer to independently parse data out of a string? For example, by using some `Hash` type that contains the actual hash bytes in binary form and has suitable `Display` and `FromStr` impls?

Reply:
The file hash isn't calculated by the implementer, as the CompilationUnit protobuf contains file hashes stored as strings.

Reply:
Perhaps it's not worth converting to binary form, but a distinct type would still be helpful for clarity and to reduce opportunity for error.

Reply:
I think that a distinct type will just introduce more overhead, because everything is done using strings. If we introduce a new type, then the caller would have to convert the string from the Compilation Unit into that type, and then this function would have to turn that back into a string.

Reply:
We're talking about a "newtype", i.e. a tuple struct with a single element, which in this case would be the wrapped string. This shouldn't introduce any overhead because it's the same actual data underneath.

Reply:
The hashes in this context will always come from sources that we trust are at least internally consistent--in particular, they're used to associate file paths to file contents coming in as part of a compile action. In that sense there's not a lot of extra signal we get from obscuring the underlying type or recording the extra invariant in the type system. If the incoming compilation records give us digests that are incorrectly formatted, the best thing we can do is try and search for the contents for those digests anyway and later complain if they can't be found.

That being said, since we're still getting to know Rust, is the newtype idiom a very strong part of Rust style? If so, it might be better to use it here, even if it means some additional ugliness at use sites (the .0 indexing, in particular).

Reply:
It's not overwhelmingly strong, no; this is definitely within the bounds of taste, as an API working in terms of `&str` would not be considered grossly unidiomatic. I do think it's helpful as it makes type signatures more self-documenting and prevents inadvertent confusion of different strings in nontrivial code, but I'm not familiar enough with the planned design here to judge how significant that's likely to be in the long run.

Reply:
These hash strings are used as keys to retrieve file data and should not leak outside a FileProvider and its (eventual) virtual file system client. I don't expect that there will be many opportunities for confusion.

Reply:
Resolving this for now. If the need arises we can always revisit this option